File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.java
class Solution {
    public int maxProfit(int[] prices) {
        // buy cell cooldown
        int n = prices.length;
        int[] buy = new int[n];
        int[] sell = new int[n];
        int[] cooldown = new int[n];

        buy[0] = -prices[0];

        for (int i=1; i<prices.length; ++i) {
            buy[i] = Math.max(buy[i-1], cooldown[i-1] - prices[i]);
            sell[i] = buy[i-1] + prices[i];
            cooldown[i] = Math.max(sell[i-1], cooldown[i-1]);  
        }

        return Math.max(sell[n-1], cooldown[n-1]);
    }
}

File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.kt
class Solution {
    fun maxProfit(prices: IntArray): Int {
        var n: Int = prices.size
        var buy: Int = -prices[0]
        var sell: Int = 0
        var cooldown: Int = 0

        for (i: Int in 1 until n) {
            var prevBuy = buy
            var prevSell = sell
            var prevCooldown = cooldown

            // 사는건 전날에 cooldown에서 
            buy = maxOf(prevCooldown - prices[i], prevBuy)
            sell = prevBuy + prices[i]
            cooldown = maxOf(prevSell, prevCooldown)
        } 
        return maxOf(sell, cooldown)
    }
}